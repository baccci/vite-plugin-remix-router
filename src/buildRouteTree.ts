import fs from 'fs'
import path from 'path'
import { ResolvedOptions } from './types'
import { isDirectory } from './utils'

export class RouteNode {
  name: string

  path: string

  children: Array<RouteNode> = []

  constructor(filePath: string) {
    this.name = path.parse(filePath).name
    this.path = filePath
  }
}

export function buildRouteTree(options: ResolvedOptions): RouteNode {
  function createNode(filePath: string) {
    const node = new RouteNode(filePath)

    if (isDirectory(absolutePath(filePath))) {
      const children = resolveChildren(filePath)

      node.children = children.map((child) =>
        createNode(`${filePath}/${child}`),
      )
    }

    return node
  }

  function absolutePath(filePath: string) {
    return path.join(options.root, filePath)
  }

  function resolveChildren(directoryPath: string) {
    const children = fs.readdirSync(absolutePath(directoryPath))

    return children.filter((child) => {
      const childPath = path.join(directoryPath, child)

      if (isDirectory(childPath)) {
        return true
      }

      const extension = path.extname(childPath)
      return options.extensions.includes(extension.substring(1))
    })
  }

  const root = createNode(options.pageDir)
  root.name = '/'

  return root
}
