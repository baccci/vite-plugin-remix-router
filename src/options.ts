import { normalizePath } from 'vite'
import { Options, ResolvedOptions, UserOptions } from './types'

const defaultOptions: Options = {
  routesDirectory: 'src/routes',
  extensions: ['tsx', 'jsx'],
}

let resolvedOptions: ResolvedOptions | null = null

export function resolveOptions(root: string, userOptions?: UserOptions) {
  return {
    root: root ?? normalizePath(process.cwd()),
    ...defaultOptions,
    ...userOptions,
  }
}

export function setOptions(options: ResolvedOptions) {
  resolvedOptions = options
}

export function getOptions() {
  if (resolvedOptions === null) {
    throw new Error('Something went wrong. Unable to resolve "UserOptions".')
  }

  return resolvedOptions
}
