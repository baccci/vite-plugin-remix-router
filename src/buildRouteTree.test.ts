import { buildRouteTree, RouteNode } from './buildRouteTree'
import { expect } from 'vitest'
import { setOptions } from './options'

setOptions({
  root: process.cwd(),
  extensions: ['tsx', 'jsx'],
  pageDir: 'example/src/pages',
})

const root = buildRouteTree()

it('should scan the pages directory and generate the routeTree', () => {
  expect(root).toBeInstanceOf(RouteNode)
  expect(root.name).toEqual('/')
  expect(root.path).toEqual('example/src/pages')
})

it('should generate children nodes of the root', () => {
  expect(root.children).toHaveLength(5)

  const children = root.children.map((child) => child.name)
  expect(children).toEqual(['$', '__auth', '__panel', 'about', 'index'])
})

it('should recursively generate children nodes for each directory', () => {
  const directories = root.children.filter((child) => child.children.length > 0)

  expect(directories).toHaveLength(2)
  const [auth, panel] = directories

  expect(auth.name).toEqual('__auth')
  expect(auth.children).toHaveLength(1)
  expect(auth.children.map((child) => child.name)).toEqual(['login'])

  expect(panel.name).toEqual('__panel')
  expect(panel.children).toHaveLength(2)
  expect(panel.children.map((child) => child.name)).toEqual(['posts', 'users'])
})

it('should exclude files which are not in the extensions list', () => {
  const ignored = root.children.find((child) => child.name === 'ignored')
  expect(ignored).toBeUndefined()

  const posts = root.children.find((child) => child.name === 'posts')

  const ignoredJson = posts?.children.find((child) => child.name === 'ignored')
  expect(ignoredJson).toBeUndefined()
})

it('should match snapshot', () => {
  expect(root).toMatchSnapshot()
})
