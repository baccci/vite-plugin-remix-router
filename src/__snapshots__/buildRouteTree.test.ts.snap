// Vitest Snapshot v1

exports[`should match snapshot 1`] = `
RouteNode {
  "children": [
    RouteNode {
      "children": [],
      "name": "$",
      "path": "example/src/pages/$.tsx",
    },
    RouteNode {
      "children": [],
      "name": "about",
      "path": "example/src/pages/about.tsx",
    },
    RouteNode {
      "children": [],
      "name": "index",
      "path": "example/src/pages/index.tsx",
    },
    RouteNode {
      "children": [
        RouteNode {
          "children": [],
          "name": "$slug",
          "path": "example/src/pages/posts/$slug.tsx",
        },
        RouteNode {
          "children": [],
          "name": "index",
          "path": "example/src/pages/posts/index.tsx",
        },
      ],
      "isDirectory": true,
      "layoutPath": undefined,
      "name": "posts",
      "path": "example/src/pages/posts",
    },
    RouteNode {
      "children": [
        RouteNode {
          "children": [
            RouteNode {
              "children": [],
              "name": "index",
              "path": "example/src/pages/users/$user/index.tsx",
            },
            RouteNode {
              "children": [],
              "name": "profile",
              "path": "example/src/pages/users/$user/profile.tsx",
            },
            RouteNode {
              "children": [],
              "name": "settings",
              "path": "example/src/pages/users/$user/settings.tsx",
            },
          ],
          "isDirectory": true,
          "layoutPath": "example/src/pages/users/$user.tsx",
          "name": "$user",
          "path": "example/src/pages/users/$user",
        },
        RouteNode {
          "children": [],
          "name": "index",
          "path": "example/src/pages/users/index.tsx",
        },
      ],
      "isDirectory": true,
      "layoutPath": undefined,
      "name": "users",
      "path": "example/src/pages/users",
    },
  ],
  "isDirectory": true,
  "layoutPath": undefined,
  "name": "/",
  "path": "example/src/pages",
}
`;
